File: Backend/src/main/java/com/etp/questforhealth/service/DoctorService.java
package com.etp.questforhealth.service;

import com.etp.questforhealth.endpoint.dto.CredentialsDto;
import com.etp.questforhealth.entity.Credentials;
import com.etp.questforhealth.entity.Doctor;
import com.etp.questforhealth.exception.InvalidLoginException;
import com.etp.questforhealth.exception.NotFoundException;
import com.etp.questforhealth.exception.ValidationException;

import java.util.List;

public interface DoctorService {

    /**
     * Checks if the login details are valid
     * @param credentials of the doctor to check the login
     * @return the whole doctor entity if the login is successful
     * @throws RuntimeException if something went wrong.
     * @throws InvalidLoginException if email and/or password are/is wrong.
     */
    Doctor checkLogin(Credentials credentials);

    /**
     * Gets all stored doctors
     * @return a list of all stored doctors
     * @throws RuntimeException if something went wrong.
     */
    List<Doctor> getAllDoctors();

    /**
     * Gets the doctor of the specific id
     * @return the doctor with the specific id
     * @throws RuntimeException if something went wrong
     * @throws NotFoundException if no doctor with that id was found.
     */
    Doctor getOneById(int id);

    /**
     * Creates a new doctor patient relationship
     * @param doctorId id of the doctor
     * @param userId id of the user to get assigned to the doctor
     * @return true if worked
     * @throws RuntimeException if something went wrong
     * @throws ValidationException if the data is not valid
     */
    boolean assignNewPatient(int doctorId, int userId);

    /**
     * Removes a doctor patient relationship
     * @param doctorId id of the doctor
     * @param userId id of the user to get assigned to the doctor
     * @return true if worked
     * @throws RuntimeException if something went wrong
     * @throws ValidationException if the data is not valid
     */
    boolean removePatient(int doctorId, int userId);
}

File: Frontend/quest-for-health/src/app/service/doctor.service.ts
import { Injectable } from '@angular/core';
import {environment} from '../../environments/environment';
import {HttpClient, HttpHeaders} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Doctor} from '../dto/doctor';
import {Credentials} from '../dto/credentials';
import {DoctorUserRelation} from '../dto/doctor-user-relation';

const baseUri = environment.backendUrl + '/doctors';

@Injectable({
  providedIn: 'root'
})
export class DoctorService {

  private httpOption = {headers: new HttpHeaders({'Content-Type': 'application/json'})};

  constructor(private httpClient: HttpClient) { }

  /**
   * Get a list of all doctors
   */
  getAllDoctors(): Observable<Doctor[]> {
    console.log('getAllDoctors()');
    return this.httpClient.get<Doctor[]>(baseUri);
  }

  /**
   * Get the information about one doctor
   * @param id of the doctor to get the information for
   */
  getDoctorById(id: number): Observable<Doctor> {
    console.log('getDoctor');
    return this.httpClient.get<Doctor>(baseUri + '/' + id);
  }

  /**
   * Checks the login data for a doctor is valid
   * @param cred Credentials (email, password) for a doctor
   */
  checkLogin(cred: Credentials): Observable<Doctor>{
    console.log('Login');
    return this.httpClient.post<Doctor>(baseUri + '/login', JSON.stringify(cred), this.httpOption);
  }

  /**
   * Assignes a new patient to the doctor
   * @param rel Doctor user relationship that should be added
   */
  assignNewPatient(rel: DoctorUserRelation): Observable<DoctorUserRelation>{
    console.log('assignNewPatient');
    return this.httpClient.post<DoctorUserRelation>(baseUri + '/relation', JSON.stringify(rel), this.httpOption);
  }

  /**
   * Removes a patient to the doctor
   * @param rel Doctor user relationship that should be removed
   */
  removePatient(rel: DoctorUserRelation): Observable<boolean>{
    console.log('assignNewPatient');
    return this.httpClient.delete<boolean>(baseUri + '/relation/' + rel.docId + '/' + rel.userId);
  }
}
